### PR TITLE
Migrate Smarty from library and into vendors, and update calls to smarty

### PR DESCRIPTION
**Pull request by Wakie87**

### CODE REVIEW

**Comment by bradymiller** [comment]
So you are aware, here's the openemr wiki page for the smarty module:
http://www.open-emr.org/wiki/index.php/Smarty_Module

I haven't looked at this in detail yet. Quick question, though. Where do the custom plugins go?

Also, what are your thoughts for always removing .gitignore and .gitattributes files from the imported packages (my git repo had some odd things happen when bringing in .gitattributes via bower imports), which I do for bower imports:
http://www.open-emr.org/wiki/index.php/Bower#Bower_Method_.28preferred.29

**Comment by Wakie87** [comment]
The custom plugins currently reside in the same place. so under /library/smarty/plugins/* 

The controller construct argument calls an array for the plugins directory, so pull in the defaults ones from the vendors folder plus the custom ones in the library. (these can be moved at a later date if needed.)

File: gacl/admin/gacl_admin.inc.php [hunk 1]
```diff
@@ -58,16 +60,6 @@
...
-//Setup the Smarty Class.
-if (file_exists($gacl_options['smarty_dir'].'/Smarty.class.php')) {
-  //this is for standard use phpgacl
-  require_once($gacl_options['smarty_dir'].'/Smarty.class.php');
-}
-else {
-  //this is for during the phpgacl install process within OpenEMR installation setup
-  require_once(dirname(__FILE__).'/../../library/Smarty/Smarty.class.php');
-}
-
```
Comment (bradymiller) [inline gacl/admin/gacl_admin.inc.php]: thank you for removing this :)

File: library/classes/Controller.class.php [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
 
-require_once(dirname(__FILE__) . "/../Smarty/Smarty.class.php");
+//require_once(dirname(__FILE__) . "/../Smarty/Smarty.class.php");
 require_once(dirname(__FILE__) . "/../formdata.inc.php");
-if (!defined('SMARTY_DIR')) {
-    define("SMARTY_DIR", dirname(__FILE__) . "/../Smarty/");
-}
+// if (!defined('SMARTY_DIR')) {
+//     define("SMARTY_DIR", dirname(__FILE__) . "/../Smarty/");
+// }
```
Comment (bradymiller) [inline library/classes/Controller.class.php]: better to remove stuff than comment it out

File: vendor/composer/autoload_classmap.php [hunk 3]
```diff
@@ -6,4 +6,7 @@
 $baseDir = dirname($vendorDir);
 
 return array(
+    'Config_File' => $vendorDir . '/smarty/smarty/libs/Config_File.class.php',
+    'Smarty' => $vendorDir . '/smarty/smarty/libs/Smarty.class.php',
+    'Smarty_Compiler' => $vendorDir . '/smarty/smarty/libs/Smarty_Compiler.class.php',
```
Comment (bradymiller) [inline vendor/composer/autoload_classmap.php]: stupid question here:
Did composer do above automatically when imported smarty?

**Comment by bradymiller** [comment]
Reviewed code which looks very nice. It doesn't look like the postnuke calendar smarty calls were adjusted. I haven't been able to test it yet; should be able to test it out tomorrow.

File: vendor/composer/autoload_classmap.php (hunk 3 above)
Comment (Wakie87) [inline vendor/composer/autoload_classmap.php]: Yes, when you run `composer install` it also sets the autoload. When you bring in new classes you can also reset the autoload by doing a `composer dump-autoload`
[https://getcomposer.org/doc/03-cli.md#dump-autoload](https://getcomposer.org/doc/03-cli.md#dump-autoload)

**Comment by bradymiller** [comment]
Testing went well. Two questions:

1. When I did the composer install (after deleting contents in vendor directory), I had the following differences in the code:
https://github.com/bradymiller/openemr/commit/750f4a07add68cb0032818139758b5d8008ef80b
(are these changes significant, especially the autoloader??)

2. Would it be ok to make it a guideline to always remove the .gitignore and .gitattributes from the packages (these files have the potential to screw up the git repository; note I do this for the bower packages).
https://github.com/bradymiller/openemr/commit/b3cd349643d6447a8bb25d2e0758bd5ae3ab3748

-brady

**Comment by bradymiller** [comment]
Actually, looks like the custom plugins haven't been incorporate in the calendar calls:
interface/main/calendar/modules/PostCalendar/pcSmarty.class.php

**Comment by bradymiller** [comment]
Disregard above comment regarding calendar plugin calls. Those are specific to the calendar and should not be modified.

**Comment by Wakie87** [comment]
If you deleted all components n the vendor directoy, including composer. It will regenerate a new class hash.  [http://alanstorm.com/composer_autoloader_part_1](http://alanstorm.com/composer_autoloader_part_1)

> Whenever Composer generates a class file, it appends a unique identifying hash string to the end of the name (1d454b6a6f83c27dc0634c7142f32df1 above). This helps Composer developers when they’re working with the autoloader code, and also has the side benefit of making sure no library developers try calling these methods on their own (since the hash isn’t easily guessable). In some ways, a generated class file with a per-generation unique name is the ultimate private method.

I also think there is a possibility we are both using different composer versions as well as different php versions. composer runs certain files depending on php versions.

**Comment by bradymiller** [comment]
Looks good. rebased you commits into 1 and brought it into the main codebase. I also updated the smarty wiki page:
http://www.open-emr.org/wiki/index.php/Smarty_Module

Thanks for yet another awesome contribution. If looking for a easy composer package to add in future, I think adodb should be simple, since is centralized in the library/sql.inc script.

**Pull request closed without merging**